mail-server/borgbackup.nix

```diff
@@ -0,0 +1,72 @@
+#  the Free Software Foundation, either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program. If not, see <http://www.gnu.org/licenses/>
+
+{ config, pkgs, lib, ... }:
+
+let
+  cfg = config.mailserver.borgbackup;
+  cronTimes = cfg.cronTimes;
+
+  methodFragment = lib.optional (cfg.compression.method != null) cfg.compression.method;  
+  autoFragment =
+    if cfg.compression.auto && cfg.compression.method == null
+    then throw "compression.method must be set when using auto."
+    else lib.optional cfg.compression.auto "auto";
+  levelFragment =
+    if cfg.compression.level != null && cfg.compression.method == null
+    then throw "compression.method must be set when using compression.level."
+    else lib.optional (cfg.compression.level != null) (toString cfg.compression.level);
+  compressionFragment = lib.concatStringsSep "," (lib.flatten [autoFragment methodFragment levelFragment]);
+  compression = lib.optionalString (compressionFragment != "") "--compression ${compressionFragment}";
+
+  encryptionFragment = cfg.encryption.method;
+  passphraseFile = lib.escapeShellArg cfg.encryption.passphraseFile;
+  passphraseFragment = lib.optionalString (cfg.encryption.method != "none")
+                         (if cfg.encryption.passphraseFile != null then ''env BORG_PASSPHRASE="$(cat ${passphraseFile})"''
+                          else throw "passphraseFile must be set when using encryption.");
+
+  locations = lib.escapeShellArgs cfg.locations;
+  name = lib.escapeShellArg cfg.name;
+
+  repoLocation = lib.escapeShellArg cfg.repoLocation;
+
+  extraInitArgs = lib.escapeShellArgs cfg.extraArgumentsForInit;
+  extraCreateArgs = lib.escapeShellArgs cfg.extraArgumentsForCreate;
+
+  cmdPreexec = lib.optionalString (cfg.cmdPreexec != null) cfg.cmdPreexec;
+  cmdPostexec = lib.optionalString (cfg.cmdPostexec != null) cfg.cmdPostexec;
+
+  borgScript = ''
+    . /etc/profile
+    ${cmdPreexec}
+    ${passphraseFragment} ${pkgs.borgbackup}/bin/borg init ${extraInitArgs} --encryption ${encryptionFragment} ${repoLocation} || true
+    ionice -c 3 nice -n19 ${passphraseFragment} ${pkgs.borgbackup}/bin/borg create ${extraCreateArgs} ${compression} ${repoLocation}::${name} ${locations}
+    ${cmdPostexec}
+  '';
+in {
+  config = lib.mkIf config.mailserver.borgbackup.enable {
+    environment.systemPackages = with pkgs; [
+      borgbackup
+    ];
+
+    systemd.services.borgbackup = {
+      description = "backup via borgbackup";
+      unitConfig.Documentation = "man:borgbackup";
+      script = borgScript;
+      startAt = cfg.startAt;
```

Comment:
This should be run as the `virtualMail` user.

Reply:
I don't think so. It running as the `virtualMail` user would mean it can only ever backup the vmail directory and other directories where `virtualMail` is allowed in. This would be sufficient for the default config but would severely limit its usefulness. No `/home` directories, no `/root`. No directories of other services.
Also if you use encryption, where would you store the encryption passphrase other than in `/root`? You want a location which is not readable for regular users. You can of course create a directory specifically for this but isn't `/root/` a natural choice for this already?
Also borg does not listen on any ports. The only time it would connect to the outside world is if you push your backups to a remote server. And that connection is at least secured by ssh.
This being said I am paranoid when it comes to servers exposed to the internet. Thus I'll look into adding an option for making borg run as a user with fewer privileges for those who want it.

Reply:
If you also want to backup other stuff, you shouldn't add this module here but in the Nixpkgs repo. Then, you could add a simple wrapper here that makes use of your NixOS module.
But you'd have to be quick for 18.03 :D

Reply:
Regarding key storage: If you only want to backup mail, giving the `virtualMail` user access to the key shouldn't be a problem, because it has access to the mails anyway :)

Reply:
If you want, I can be the one who creates a PR on the Nixpkgs side.

Reply:
I did in fact write this module with nixpkgs in mind. Adding it to nixpkgs seems a bit more involved than adding it here and I am a bit short on time at the moment.

Reply:
So, you're okay with me taking your work over to Nixpkgs?